File: rxhttp-compiler/src/main/java/com/rxhttp/compiler/DomainAnnotatedClass.java
package com.rxhttp.compiler;


import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.MethodSpec;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

import javax.lang.model.element.Modifier;
import javax.lang.model.element.VariableElement;

import rxhttp.wrapper.annotation.Domain;

public class DomainAnnotatedClass {

    private Map<String, VariableElement> mElementMap;

    public DomainAnnotatedClass() {
        mElementMap = new LinkedHashMap<>();
    }

    public void add(VariableElement variableElement) {
        Domain annotation = variableElement.getAnnotation(Domain.class);
        String name = annotation.name();
        if (name.length() <= 0) {
            name = variableElement.getSimpleName().toString();
        }
        mElementMap.put(name, variableElement);
    }

    public List<MethodSpec> getMethodList() {
        List<MethodSpec> methodList = new ArrayList<>();
        MethodSpec.Builder method;
        for (Entry<String, VariableElement> item : mElementMap.entrySet()) {
            method = MethodSpec.methodBuilder("setDomainTo" + item.getKey() + "IfAbsent")
                .addModifiers(Modifier.PUBLIC)
                .addStatement("String newUrl = addDomainIfAbsent(param.getSimpleUrl(), $T.$L)",
                    ClassName.get(item.getValue().getEnclosingElement().asType()),
                    item.getValue().getSimpleName().toString())
                .addStatement("param.setUrl(newUrl)")
                .addStatement("return (R)this")
                .returns(RxHttpGenerator.r);
            methodList.add(method.build());
        }

        //对url添加域名方法
        method = MethodSpec.methodBuilder("addDomainIfAbsent")
            .addModifiers(Modifier.PRIVATE, Modifier.STATIC)
            .addParameter(String.class, "url")
            .addParameter(String.class, "domain")
            .addCode("if (url.startsWith(\"http\")) return url;\n" +
                "if (url.startsWith(\"/\")) {\n" +
                "    if (domain.endsWith(\"/\"))\n" +
                "        return domain + url.substring(1);\n" +
                "    else\n" +
                "        return domain + url;\n" +
                "} else if (domain.endsWith(\"/\")) {\n" +
                "    return domain + url;\n" +
                "} else {\n" +
                "    return domain + \"/\" + url;\n" +
                "}")
            .returns(String.class);
        methodList.add(method.build());
        return methodList;
    }
}

File: rxhttp-compiler/src/main/java/com/rxhttp/compiler/DomainAnnotatedClass.kt
package com.rxhttp.compiler

import com.squareup.javapoet.ClassName
import com.squareup.javapoet.MethodSpec
import rxhttp.wrapper.annotation.Domain
import java.util.*
import javax.lang.model.element.Modifier
import javax.lang.model.element.VariableElement

class DomainAnnotatedClass {

    private val mElementMap = LinkedHashMap<String, VariableElement>()

    fun add(variableElement: VariableElement) {
        val annotation = variableElement.getAnnotation(Domain::class.java)
        var name: String = annotation.name
        if (name.isEmpty()) {
            name = variableElement.simpleName.toString()
        }
        mElementMap[name] = variableElement
    }

    //对url添加域名方法
    val methodList: List<MethodSpec>
        get() {
            val methodList = ArrayList<MethodSpec>()
            for ((key, value) in mElementMap) {
                methodList.add(MethodSpec.methodBuilder("setDomainTo" + key + "IfAbsent")
                    .addModifiers(Modifier.PUBLIC)
                    .addStatement("String newUrl = addDomainIfAbsent(param.getSimpleUrl(), \$T.\$L)",
                        ClassName.get(value.enclosingElement.asType()),
                        value.simpleName.toString())
                    .addStatement("param.setUrl(newUrl)")
                    .addStatement("return (R)this")
                    .returns(r).build())
            }

            //对url添加域名方法
            methodList.add(
                MethodSpec.methodBuilder("addDomainIfAbsent")
                    .addModifiers(Modifier.PRIVATE, Modifier.STATIC)
                    .addParameter(String::class.java, "url")
                    .addParameter(String::class.java, "domain")
                    .addCode("""
                         if (url.startsWith("http")) return url;
                         if (url.startsWith("/")) {
                             if (domain.endsWith("/"))
                                 return domain + url.substring(1);
                             else
                                 return domain + url;
                         } else if (domain.endsWith("/")) {
                             return domain + url;
                         } else {
                             return domain + "/" + url;
                         }
                    """.trimIndent())
                    .returns(String::class.java).build())
            return methodList
        }
}